parser/impl/parser.cpp

```diff
@@ -469,17 +469,163 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
 // ---- @aabounegm
 
 sPtr<ast::Expression> Parser::parseExpression() {
-    // TODO: Concrete implementation
+    return parseBinaryExpression();
+}
+
+sPtr<ast::Expression> Parser::parseUnaryExpression() {
+    Token tok = m_lexer.Peek();
+    ast::UnaryExpression expr;
+    if (opPrec(tok.type) >= 0) {
+        // if has operations on primary
+        if (tok.type == TokenType::Not || tok.type == TokenType::Sub ||
+            tok.type == TokenType::Add) {
+            // math unary operations
+            tok = m_lexer.Next();
+            expr.operation = tok;
+            expr.operand = parseUnaryExpression();
+            return std::make_shared<ast::UnaryExpression>(expr);
+
+        } else if (tok.type == TokenType::OpenParen) {
+            // parenthesis -> more priority
+            tok = m_lexer.Next();
+            sPtr<ast::Expression> expr = parseBinaryExpression(0);
+            if (m_lexer.Peek().type != TokenType::CloseParen) {
+                m_errors.push_back(Error{
+                    .pos = tok.pos,
+                    .message = "Expected to find ')'",
+                });
+                while (m_lexer.Peek().type != TokenType::CloseParen)
+                    m_lexer.Next();
+                return nullptr;
+            }
+            tok = m_lexer.Next(); // read ')'
+            return expr;
+        } else {
+            // throw error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find unary operator",
+            });
+            return nullptr;
+        }
+    } else if (isPrimitive(tok.type)) {
+        // if is pure primitive
+        tok = m_lexer.Next();
+        // check if More Then Identifier
+        if (tok.type == TokenType::Identifier) {
+            Token t = m_lexer.Peek();
+            switch (t.type) {
+            case TokenType::OpenParen:
+                return parseRoutineCall(tok);
+            case TokenType::Dot:
+            case TokenType::OpenBrack:
+                return parseModifiablePrimary(tok);
+            default:
+                break;
+            }
+        }
+        // return Primitive only
+        ast::Primitive prim;
+        prim.value = tok;
+        return std::make_shared<ast::Primitive>(prim);
+    }
     return nullptr;
 }
 
+sPtr<ast::Expression> Parser::parseBinaryExpression(int prec1) {
```

Comment:
```suggestion
sPtr<ast::Expression> Parser::parseBinaryExpression(int prec1 = 0) {
```
Please, set the prec to the default value of 0. This will make the intent of the code at the calling site (in the parse unary expression) more clear.

Reply:
Such change will result in compilation error due to the fact that I have already added this thing to the header file.